File: 39- Longest common substring/Python/solution4.py
# By using a suffix tree:

# Time complexity: O(n + m)
# Space complexity: O((n+m)²)

def getLcsFromSuffixTree(suffixTree):
  if not suffixTree.hasSuffixFromStr1 or not suffixTree.hasSuffixFromStr2:
    return 0
  else:
    maxFromChild = 0
    for child in suffixTree.children:
      maxFromChild = max(maxFromChild, getLcsFromSuffixTree(child))
    return len(suffixTree.text) + maxFromChild

def lcs(str1, str2):
  concatenation = str1 + '*' + str2 + '$'
  suffixTree = SuffixTree(concatenation)
  return getLcsFromSuffixTree(suffixTree)

File: 39- Longest common substring/JavaScript/solution4.js
// By using a suffix tree:

// Time complexity: O(n + m)
// Space complexity: O((n+m)²)

function getLcsFromSuffixTree(suffixTree){
  if(!suffixTree.hasSuffixFromStr1 || !suffixTree.hasSuffixFromStr2){
    return 0;
  }else{
    let maxFromChild = 0;
    for(let child of suffixTree.children){
      maxFromChild = Math.max(maxFromChild, getLcsFromSuffixTree(child));
    }
    return suffixTree.text.length + maxFromChild;
  }
}

function lcs(str1, str2){
  let concatenation = str1 + '*' + str2 + '$';
  let suffixTree = new SuffixTree(concatenation);
  return getLcsFromSuffixTree(suffixTree);
}


